src/protocol/prf_sharding/feature_label_dot_product.rs

```diff
@@ -0,0 +1,469 @@
+        } else {
+            let ctx_for_row_number = root_ctx
+                .narrow(&UserNthRowStep::from(row_number))
+                .set_total_records(*num_users_having_that_row_number);
+            context_per_row_depth.push(ctx_for_row_number);
+        }
+    }
+    context_per_row_depth
+}
+
+fn chunk_rows_by_user<FV>(
+    input_rows: Vec<PrfShardedIpaInputRow<FV>>,
+) -> Vec<Vec<PrfShardedIpaInputRow<FV>>>
+where
+    FV: GaloisField,
+{
+    let mut rows_for_user: Vec<PrfShardedIpaInputRow<FV>> = vec![];
+
+    let mut rows_chunked_by_user = vec![];
+    for row in input_rows {
+        if rows_for_user.is_empty() || row.prf_of_match_key == rows_for_user[0].prf_of_match_key {
+            rows_for_user.push(row);
+        } else {
+            rows_chunked_by_user.push(rows_for_user);
+            rows_for_user = vec![row];
+        }
+    }
+    if !rows_for_user.is_empty() {
+        rows_chunked_by_user.push(rows_for_user);
+    }
+
+    rows_chunked_by_user
+}
+
+/// Sub-protocol of the PRF-sharded IPA Protocol
+///
+/// After the computation of the per-user PRF, addition of dummy records and shuffling,
+/// the PRF column can be revealed. After that, all of the records corresponding to a single
+/// device can be processed together.
+///
+/// This circuit expects to receive records from multiple users,
+/// but with all of the records from a given user adjacent to one another, and in time order.
+///
+/// This circuit will compute attribution, and per-user capping.
+///
+/// The output of this circuit is the input to the next stage: Aggregation.
+///
+/// # Errors
+/// Propagates errors from multiplications
+/// # Panics
+/// Propagates errors from multiplications
+pub async fn compute_feature_label_dot_product<C, FV, F, S>(
+    sh_ctx: C,
+    input_rows: Vec<PrfShardedIpaInputRow<FV>>,
+) -> Result<Vec<S>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + Serializable + SecureMul<C::UpgradedContext<F>>,
+    FV: GaloisField,
+    F: PrimeField + ExtendableField,
+{
+    assert!(FV::BITS > 0);
+
+    let rows_chunked_by_user = chunk_rows_by_user(input_rows);
+    let histogram = compute_histogram_of_users_with_row_count(&rows_chunked_by_user);
+    let binary_validator = sh_ctx.narrow(&Step::BinaryValidator).validator::<Gf2>();
+    let binary_m_ctx = binary_validator.context();
+    let mut num_users_who_encountered_row_depth = Vec::with_capacity(histogram.len());
+    let ctx_for_row_number = set_up_contexts(&binary_m_ctx, &histogram);
+    let mut futures = Vec::with_capacity(rows_chunked_by_user.len());
+    for rows_for_user in rows_chunked_by_user {
+        for i in 0..rows_for_user.len() {
+            if i >= num_users_who_encountered_row_depth.len() {
+                num_users_who_encountered_row_depth.push(0);
+            }
```

Comment:
Rather than adding one-by-one, you can do:

```rust
if num_users_who_encountered_row_depth.len() < rows_for_user.len() {
  num_users_who_encountered_row_depth.resize(rows_for_user.len(), 0);
}
```

Reply:
Thanks for telling me about this trick! I've used it.